### DISCUSSION TITLE
Lit-localize is confusing me.

### DISCUSSION THREAD

**Discussion by kensternberg-authentik** [lit/lit#4489] [Q&A]
I've been working on the [authentik SSO](https://github.com/goauthentik/authentik) project, and we've been using lit-localize for about a year now in runtime mode, quite happily and successfully.  Today, though, I hit something... bewildering.

In `tsconfig.json`, after changing from `"module": "esnext", "moduleResolution": "node"` to `"module": "nodenext", "moduleResolution": "nodenext"`, `lit-localize extract` suddenly began reporting:

```
Placeholder error in pl localization of sc9175cb129fdc306: unexpected "${item.verboseName}"
Placeholder error in pl localization of sc9175cb129fdc306: missing "${item.policyObj?.name}"
```

But if I go look at the Polish file, I don't see any conflict.

```
      <trans-unit id="sc9175cb129fdc306">
        <source>Update <x id="0" equiv-text="${item.verboseName}"/></source>
        <target>Zaktualizuj <x id="0" equiv-text="${item.verboseName}"/></target>
      </trans-unit>
```

The documentation says that "the code inside the expression does NOT affect the ID of the translation unit, only its position."  

First, I take from this that, even if I changed the variable name, the translations files ought not to need manual intervention. That there's an expression in that position ought to be enough, right?

Second, why the heck would changing the module resolution mode eff up lit-localize this way?

I don't want to mark this as a bug (yet); I don't know what's causing it, I don't have a solid repro, and it may just be all on my end, but, like, what the heck?

**Comment by augustjk** [comment]
While code inside expression does not affect the ID generation, in cases of nested templates, it can affect `equiv-text` which I believe does need to match. So if the variable name is changed, the source can need updating.

However, changing those tsconfig compile options alone shouldn't really change things unless it resulted in a different source code being loaded due to the different module resolution mode.

Based on the message, it seems like it found `${item.policyObj?.name}` in the code without seeing `${item.verboseName}` so it seems the source changed?

**Comment by kensternberg-authentik** [comment]
I don't understand.  Does that mean I need a different translation unit for each of these otherwise identical tagged strings?

```
<h1 class="pf-c-title pf-m-2xl">${msg(str`Update ${this.objectLabel}`)}</h1>
<span slot="header"> ${msg(str`Update ${item.policyObj?.name}`)} </span>
<span slot="header"> ${msg(str`Update ${item.stageObj?.verboseName}`)} </span>
<span slot="header"> ${msg(str`Update ${item.verboseName}`)} </span>
```

**Reply by augustjk** [reply]
That is an excellent point that I overlooked. You're absolutely right. They should all result in a single trans-unit with the same id and equiv-text then shouldn't matter.

Reading back at your original post, I agree with your point:
> even if I changed the variable name, the translations files ought not to need manual intervention

So then this might be a bug in the extraction analysis, and my guess is the changing module resolution mode might've changed the order in which the translation targets were encountered in the source which resulted in the discrepancy.

**Reply by augustjk** [reply]
Thank you for raising this question/issue. I've filed https://github.com/lit/lit/issues/4502
I need to look closer at exactly what the validation is doing and whether it is actually okay to ignore the `equiv-text`.
